### PR TITLE
BUG make sure to handle zipped keys properly in reordering

### LINKED ISSUE #1847: BUG: conda-smithy v3.31 mismatches cuda_compiler & image, breaks zip_keys
I'm getting the following rerender in https://github.com/conda-forge/arrow-cpp-feedstock/pull/1325, which mismatches the cuda_compiler with the image, and then predictably fails.
```diff
--- a/.azure-pipelines/azure-pipelines-linux.yml
+++ b/.azure-pipelines/azure-pipelines-linux.yml
@@ -11,7 +11,7 @@ jobs:
       linux_64_cuda_compiler_version11.2:                          # CUDA
         CONFIG: linux_64_cuda_compiler_version11.2
         UPLOAD_PACKAGES: 'True'
-        DOCKER_IMAGE: quay.io/condaforge/linux-anvil-cuda:11.2     # CUDA
+        DOCKER_IMAGE: quay.io/condaforge/linux-anvil-cos7-x86_64   # !! no CUDA !!
       linux_64_cuda_compiler_versionNone:
         CONFIG: linux_64_cuda_compiler_versionNone
         UPLOAD_PACKAGES: 'True'
--- a/.ci_support/linux_64_cuda_compiler_version11.2.yaml
+++ b/.ci_support/linux_64_cuda_compiler_version11.2.yaml           # CUDA
@@ -7,15 +7,15 @@ bzip2:
 cuda_compiler:
-- nvcc                                                             # CUDA
+- None                                                             # !! no CUDA !!
 cuda_compiler_version:
 - '11.2'                                                           # CUDA
 cuda_compiler_version_min:
@@ -23,23 +23,23 @@ cuda_compiler_version_min:
 docker_image:
-- quay.io/condaforge/linux-anvil-cuda:11.2
+- quay.io/condaforge/linux-anvil-cos7-x86_64
 gflags:
--- a/.ci_support/linux_64_cuda_compiler_versionNone.yaml
+++ b/.ci_support/linux_64_cuda_compiler_versionNone.yaml           # no CUDA
@@ -9,13 +9,13 @@ c_compiler:
 cuda_compiler:
-- None                                                             # no CUDA
+- nvcc                                                             # !! CUDA !!
 cuda_compiler_version:
 - None                                                             # no CUDA
 cuda_compiler_version_min:
```

In fact, for 4/4 builds on the arrow feedstock, the rerender randomly mismatches at least two jobs, seemingly by messing up the zip_keys. Not sure what's happening here or whether #1815 had a bug or uncovered another problem.

@mbargull @beckermr 
CC @conda-forge/core

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [x] Added a ``news`` entry
* [x] add tests
* [x] put code in separate function possibly

closes #1847

### CODE REVIEW

**Comment by mbargull** [comment]
> ... but needs specific tests

This one is a tiniest I could come up with:
```yaml
# conda_build_config.yaml
a:
  - 1
  - 2
  - 2
b:
  - 1
  - 2
  - 1

zip_keys:
  -
    - a
    - b
```
```yaml
# meta.yaml
package:
  name: test_skip_some_zip_keys_combination
  version: 1.0

build:
  skip: true  # [not linux64 or a != b]

about: {}
```

i.e., the issue came up with the `arrow-cpp` feedstock since it has a
```yaml
  skip: true  # [cuda_compiler_version not in ("None", cuda_compiler_version_min)]
```
which skips the `cuda_compiler_version: "11.8"` variant.

> and some cleanup possibly

Just putting this partially in a aptly named function would possibly suffice :shrug:.

File: conda_smithy/configure_feedstock.py [hunk 1]
```diff
@@ -383,26 +403,119 @@ def _collapse_subpackage_variants(
...
+    for k, v in squished_used_variants.items():
+        if k in all_used_vars and k not in zipped_keys:
             used_key_values[k] = v
 
+    logger.debug(
+        "post input reorder used_key_values {}".format(
+            pprint.pformat(used_key_values)
+        )
+    )
+
```
Comment (mbargull) [inline conda_smithy/configure_feedstock.py:518]:
```python
def _get_used_key_values_by_input_order(
    squished_input_variants,
    squished_used_variants,
    all_used_vars,
):
...
    used_key_values = _get_used_key_values_by_input_order(...)
```
or something like that?

Comment (mbargull) [reply conda_smithy/configure_feedstock.py:518]: or, I guess, just `_get_used_key_values`

**Comment by beckermr** [comment]
This PR is ready for review @mbargull!

**Review by mbargull — APPROVED** [review]
Thanks!!

**Comment by mbargull** [comment]
Can also confirm this working on `arrow-cpp`... it just took nearly 30 minutes to render O__O .

**Comment by beckermr** [comment]
The big test case in the code was pulled directly from the failing bits of the arrow cpp internals fyi.

**Comment by jakirkham** [comment]
Wow! Am impressed you were both able to figure this out so quickly

The additional logging should help if we need to do a deep dive here again

Thank you both for your hard work here! 🙏

**Pull request merged**